application/app/services/detached_process_controller.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+# DetachedProcessController: orchestrates multiple ProcessExecutors
+class DetachedProcessController
+  include LoggingCommon
+
+  def initialize(services, interval: Configuration.detached_controller_interval)
+    @interval = interval
+    @executor_threads = services.map { |service| ProcessExecutor.new(service) }.to_h { |executor| [executor, nil] }
+  end
+
+  def run
+    log_info('Process controller started', {executors: @executor_threads.size})
+
+    loop do
+      @executor_threads.each do |executor, thread|
+        if thread.nil? || !thread.alive?
+          log_info('Launching executor', {name: executor.name})
+          @executor_threads[executor] = executor.start
+        end
+      end
+
+      sleep @interval
+
+      all_idle = @executor_threads.values.all? { |t| !t.alive? }
+
+      if all_idle
+        log_info('All executors idle. Shutting down...', {executors: @executor_threads.size})
+        @executor_threads.each_key(&:shutdown)
+        log_info('Completed', {executors: @executor_threads.size})
+        break
+      end
```

Comment:
beware of `Thread.alive?`. According to the documentation https://ruby-doc.org/3.1.6/Thread.html#method-i-alive-3F it returns false if is sleeping and true if it is running. The thread can be in memory without having finished and get to sleep for a while due to a I/O temporal blockage or because the processor is running another thread. So we shouldn't use `alive?` for this purpose.

Personally I prefer to rely on `Thread.join` when I work with threads and I encourage you to use it in this code. Alternatively, you may consider checking `Thread.status` as `false` or `nil` as is documented here https://ruby-doc.org/3.1.6/Thread.html#method-i-status